core/src/providers/anthropic.rs

```diff
@@ -114,22 +195,118 @@ impl TryFrom<&ChatMessage> for AnthropicChatMessage {
     }
 }
 
+// Tools.
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct AnthropicTool {
+    pub name: String,
+    pub description: Option<String>,
+    pub input_schema: Option<Value>,
+}
+
+impl TryFrom<&ChatFunction> for AnthropicTool {
+    type Error = anyhow::Error;
+
+    fn try_from(f: &ChatFunction) -> Result<Self, Self::Error> {
+        Ok(AnthropicTool {
+            name: f.name.clone(),
+            description: f.description.clone(),
+            input_schema: f.parameters.clone(),
+        })
+    }
+}
+
 #[derive(Serialize, Deserialize, Debug, Clone)]
 pub struct Usage {
     pub input_tokens: u64,
     pub output_tokens: u64,
 }
 
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct AnthropicStreamContent {
+    pub r#type: String,
+    pub text: String,
+}
+
 #[derive(Serialize, Deserialize, Debug, Clone)]
-pub struct ChatResponse {
+pub struct StreamChatResponse {
     pub id: String,
     pub model: String,
     pub role: AnthropicChatMessageRole,
-    pub content: Vec<AnthropicContent>,
+    pub content: Vec<AnthropicStreamContent>,
     pub stop_reason: Option<StopReason>,
     pub usage: Usage,
 }
 
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct ChatResponse {
+    id: String,
+    model: String,
+    role: AnthropicChatMessageRole,
+    content: Vec<AnthropicResponseContent>,
+    stop_reason: Option<StopReason>,
+    usage: Usage,
+}
+
+impl TryFrom<StreamChatResponse> for ChatResponse {
+    type Error = anyhow::Error;
+
+    fn try_from(cr: StreamChatResponse) -> Result<Self, Self::Error> {
+        let content = cr
+            .content
+            .into_iter()
+            .map(AnthropicResponseContent::try_from)
+            .collect::<Result<Vec<AnthropicResponseContent>, anyhow::Error>>()?;
+
+        Ok(ChatResponse {
+            id: cr.id.clone(),
+            model: cr.model.clone(),
+            role: cr.role.clone(),
+            content,
+            stop_reason: cr.stop_reason.clone(),
+            usage: cr.usage.clone(),
+        })
+    }
+}
+
+// This code converts a ChatResponse to a ChatMessage, but only supports one tool call.
+// It takes the first tool call from the vector of AnthropicResponseContent,
+// potentially discarding others. Anthropic often returns the CoT content as a first message,
+// which gets combined with the first tool call in the resulting ChatMessage.
+impl TryFrom<ChatResponse> for ChatMessage {
+    type Error = anyhow::Error;
+
+    fn try_from(cr: ChatResponse) -> Result<Self, Self::Error> {
+        let text_content = cr.content.iter().find_map(|item| match item.get_text() {
+            Some(text) => Some(text.clone()),
```

Comment:
do you need to clone here ? Can you clone closer to when you actually need it so it's more obvious why ?